src/codelabs/iterables.md

```diff
@@ -0,0 +1,608 @@
+
+* As an **expression**: The test code has one line that uses arrow syntax (`=>`).
+* As a **block**: The test code has multiple lines between brackets and a return statement.
+* As a **function**: The test code is in an external function that's passed to
+the `firstWhere()` method as a parameter.
+
+There is no right or wrong way. Use the way that works best for you, 
+and makes your code easier to read and understand.
+
+In the example, `firstWhereWithOrElse()` calls `firstWhere()` with the optional named
+parameter `orElse`, which provides an alternative when an element isn't found.
+In this case, the text `'None!'` is returned because no element satisfies the provided
+condition.
+
+{{site.alert.note}}
+  If no element satisfies the test predicate and
+  the `orElse` parameter isn't provided,
+  then `firstWhere()` throws a [StateError][StateError class].
+{{site.alert.end}}
+
+{{site.alert.secondary}}
+  **Quick review:**
+  * The elements of an `Iterable` must be accessed sequentially.
+  * The easiest way to iterate through all the elements is using a `for-in` loop.
+  * You can use the `first` and `last` getters to get the first and last elements.
+  * You can also find the first element that matches a condition with `firstWhere()`.
+  * You can write test predicates as expressions, blocks, or functions.
+
+**Key terms:**
+* **Predicate**: A function that returns `true` when a certain condition is satisfied.
+{{site.alert.end}}
+
+### Exercise: Practice writing a test predicate
+
+The following exercise is a failing unit test that contains a partially complete code snippet.
+Your task is to complete the exercise by writing code to make the tests pass.
+You don't need to implement `main()`.
+
+This exercise introduces `singleWhere()`. This method works similarly to `firstWhere()`, but
+in this case, it expects only one element of the `Iterable` to match the predicate.
+If more than one or no element in the `Iterable` satisfies the predicate condition, then the
+method throws a [StateError][StateError class] exception.
+
+{{site.alert.warn}}
+  `singleWhere()` steps through the whole `Iterable` until the last element, which can cause
+  problems if the `Iterable` is infinite or contains a large collection of elements.
+{{site.alert.end}}
+
+Your goal is to implement the predicate for `singleWhere()` that satisfies the following conditions:
+
+* The element contains the character `'a'`.
+* The element starts with the character `'M'`.
+
+All the elements in the test data are [strings;][String class]
+you can check the class documentation for help.
+
+[//]: https://gist.github.com/241c6bc591f9436a9be0116724222953
+<iframe
+  src="{{site.dartpad-embed}}?id=241c6bc591f9436a9be0116724222953&theme=dark"
+  frameborder="no"
+  height="250"
+  width="100%">
+</iframe>
+
+## Checking conditions
+
+When working with `Iterable`, sometimes you need to verify that the elements of a collection
+satisfy some condition.
```

Comment:
This was a confusing sentence to read after having just finished an exercise about verifying that an element satisfies a condition (and the prior section about finding the first element that satisfies a condition).

I'd recommend rewording this to acknowledge that, while we have reviewed some use cases for predicates, there are new ones to cover.